trio/_core/_multierror.py

```diff
@@ -218,33 +236,54 @@ def __new__(cls, exceptions, *, _collapse=True):
             # In an earlier version of the code, we didn't define __init__ and
             # simply set the `exceptions` attribute directly on the new object.
             # However, linters expect attributes to be initialized in __init__.
+            from_class: type[Self] | type[NonBaseMultiError] = cls
             if all(isinstance(exc, Exception) for exc in exceptions):
-                cls = NonBaseMultiError
+                from_class = NonBaseMultiError
 
-            return super().__new__(cls, "multiple tasks failed", exceptions)
+            # Ignoring arg-type: 'Argument 3 to "__new__" of "BaseExceptionGroup" has incompatible type "list[BaseException]"; expected "Sequence[_BaseExceptionT_co]"'
+            # We have checked that exceptions is indeed a list of BaseException objects, this is fine.
+            new_obj = super().__new__(from_class, "multiple tasks failed", exceptions)  # type: ignore[arg-type]
+            assert isinstance(new_obj, (cls, NonBaseMultiError))
+            return new_obj
 
-    def __reduce__(self):
+    def __reduce__(
+        self,
+    ) -> tuple[object, tuple[type[Self], list[BaseException]], dict[str, bool],]:
         return (
             self.__new__,
             (self.__class__, list(self.exceptions)),
             {"collapse": self.collapse},
         )
 
-    def __str__(self):
+    def __str__(self) -> str:
         return ", ".join(repr(exc) for exc in self.exceptions)
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return f"<MultiError: {self}>"
 
-    def derive(self, __excs):
+    @overload
+    def derive(self, __excs: Sequence[Exception]) -> NonBaseMultiError:
+        ...
+
+    @overload
+    def derive(self, __excs: Sequence[BaseException]) -> MultiError:
+        ...
+
+    def derive(
+        self, __excs: Sequence[Exception | BaseException]
```

Comment:
We dropped 3.7, we can use `/` now (even though IK it was like `__` before, might as well clean it up).

Reply:
Changed in [5ad1cab](https://github.com/python-trio/trio/pull/2742/commits/5ad1cab84968e3ce6384e95abb433802e0448c43)